File: 4. Median of Two Sorted Arrays/Sort and Pick/median_of_two_sorted_arrays.kt
fun get_median (list: Array<Int>): Float{
	val len_list: Int = list.size;
	val even_length: Boolean = (len_list % 2 == 0);
	var median: Float = 0.0f;
	
	if (even_length){
		median = ((list[len_list / 2].toFloat() + list[(len_list / 2) - 1].toFloat()) / 2);
	} else {
		median = (list[len_list / 2]).toFloat();
	}
	
	return (median);
}

fun combine_list (list1_original: Array<Int>, list2_original: Array<Int>) : Array<Int>{
	var list1: Array<Int> = list1_original;
	var list2: Array<Int> = list2_original;
	
	list1.sort();
	list2.sort();
	
	val len_list1: Int = list1.size;
	val len_list2: Int = list2.size;
	
	var nje: Array<Int> = Array<Int>(len_list1 + len_list2){0};
	
	var nje_cursor: Int = 0;
	var list1_cursor: Int = 0;
	var list2_cursor: Int = 0;
	
	while ((list1_cursor < len_list1) || (list2_cursor < len_list2)){
		var element_list1: Int? = null;
		var element_list2: Int? = null;
		var appending_element: Int? = null;
		
		if (list1_cursor < len_list1){
			element_list1 = list1[list1_cursor];
		}
		
		if (list2_cursor < len_list2){
			element_list2 = list2[list2_cursor];
		}
		
		if ((element_list1 != null) && (element_list2 != null)){
			if (element_list1 < element_list2){
				appending_element = element_list1;
				list1_cursor += 1;
			} else {
				appending_element = element_list2;
				list2_cursor += 1;
			}
		} else if (element_list1 == null){
			appending_element = element_list2;
			list2_cursor += 1;
		} else if (element_list2 == null){
			appending_element = element_list1;
			list1_cursor += 1;
		}
		
		nje[nje_cursor] = appending_element!!;
		nje_cursor += 1;
	}
	
	return (nje);
}

fun main (){
	val list1: Array<Int> = arrayOf(1,3);
	val list2: Array<Int> = arrayOf(2,4);
	
	var list_combined = combine_list(list1, list2);
	
	var median = get_median(list_combined);
	
	println("This is the handy work of Kotlin.\n\n\n");
	println(median);
	println("\n\n\n");
}

File: 4. Median of Two Sorted Arrays/Sort and Pick/median_of_two_sorted_arrays.js
function toInteger (number){
	return (number - (number % 1));
}

function getMedian (list){
	let lenList = list.length;
	let evenLength = (lenList % 2 == 0);
	let median = 0.0;
	
	if (evenLength){
		median = (list[toInteger(lenList / 2)] + list[toInteger(lenList / 2) - 1]) / 2; 
	} else {
		median = (list[toInteger(lenList / 2)]);
	}
	
	return (median);
}

function combineList (list1Original, list2Original){
	let list1 = list1Original.sort(function (a, b){return a - b});
	let list2 = list2Original.sort(function (a, b){return a - b});
	
	let lenList1 = list1.length;
	let lenList2 = list2.length;
	
	let list1Cursor = 0;
	let list2Cursor = 0;
	
	let nje = [];
	
	while ((list1Cursor < lenList1) || (list2Cursor < lenList2)){
		let elementList1 = null;
		let elementList2 = null;
		let appendingElement = null;
		
		if (list1Cursor < lenList1){
			elementList1 = list1[list1Cursor];
		}
		
		if (list2Cursor < lenList2){
			elementList2 = list2[list2Cursor];
		}
		
		if ((elementList1 != null) && (elementList2 != null)){
			if (elementList1 < elementList2){
				appendingElement = elementList1;
				list1Cursor += 1;
			} else {
				appendingElement = elementList2;
				list2Cursor += 1;
			}
		} else if (elementList1 == null){
			appendingElement = elementList2;
			list2Cursor += 1;
		} else if (elementList2 == null){
			appendingElement = elementList1;
			list1Cursor += 1;
		}
		
		nje.push(appendingElement);
	}
	
	return (nje);
}

let list1 = [1,3,5];
let list2 = [2,4];

let list_combined = combineList(list1, list2)

let median = getMedian(list_combined);

console.log("This is the handy work of JavaScript\n\n");
console.log(median);
console.log("\n\n\n");
